src/compiler/checker.ts

```diff
@@ -12019,47 +11982,99 @@ namespace ts {
             return getSignaturesOfStructuredType(getReducedApparentType(type), kind);
         }
 
-        function getIndexInfoOfStructuredType(type: Type, kind: IndexKind): IndexInfo | undefined {
+        function findIndexInfo(indexInfos: readonly IndexInfo[], keyType: Type) {
+            return find(indexInfos, info => info.keyType === keyType);
+        }
+
+        function findApplicableIndexInfo(indexInfos: readonly IndexInfo[], keyType: Type) {
+            // Index signatures for type 'string' are considered only when no other index signatures apply.
+            let stringIndexInfo: IndexInfo | undefined;
+            let applicableInfo: IndexInfo | undefined;
+            let applicableInfos: IndexInfo[] | undefined;
+            for (const info of indexInfos) {
+                if (info.keyType === stringType) {
+                    stringIndexInfo = info;
+                }
+                else if (isApplicableIndexType(keyType, info.keyType)) {
+                    if (!applicableInfo) {
+                        applicableInfo = info;
+                    }
+                    else {
+                        (applicableInfos || (applicableInfos = [applicableInfo])).push(info);
+                    }
+                }
+            }
+            // When more than one index signature is applicable we create a synthetic IndexInfo. Instead of computing
+            // the intersected key type, we just use unknownType for the key type as nothing actually depends on the
+            // keyType property of the returned IndexInfo.
+            return applicableInfos ? createIndexInfo(unknownType, getIntersectionType(map(applicableInfos, info => info.type)),

```

Comment:
`unknown` isn't assignable to `string | number | symbol` - maybe it'd be better to use `any` here on the offchance it leaks somewhere in the future?

Reply:
To ensure property round-tripping, our guiding principle should be what happens when you feed a particular type through a mapped type. Currently, feeding an `any` through a mapped type produces a `string` index signature. That used to make sense because it was the widest possible index signature, but you could argue we should equate `any` to `string | number | symbol` and have it be shorthand for all three index signatures. That may or may not affect existing code, but we could try.

Reply:
Oops, comment above was meant for your general question about `any` and `never`.

Reply:
Meanwhile, sure it might be better to use `any` for the key type for synthetic index infos.

Reply:
> but you could argue we should equate any to string | number | symbol and have it be shorthand for all three index signatures. That may or may not affect existing code, but we could try.

Yeah, it makes sense, imo, for a mapped type over `any` to make an `any` index signature, since that'll actually handle `symbol`s. (Just like the `any` input could have)